### PR TITLE
Add support for broadcasting over multi-dimensional inputs in KerasLayer and TorchLayer

### PR DESCRIPTION
**Pull request by DanielPolatajko**
**Context:**
Most layers in Pytorch or Keras accept arbitrary dimension inputs, where each dimension barring the last (in the case where the actual weight function of the layer operates on one-dimensional vectors) is broadcast over. This is not currently possible with Pennylane's KerasLayer or TorchLayer, where inputs must have the form (batch size, input dim) with a returned shaped of (batch size, output dim).

**Description of the Change:**
This PR adds support for broadcasting over multiple dimensions of inputs to TorchLayer and KerasLayer, not just the batch. That is, inputs of shape (batch size, d_n, ..., d_1, d_0) will return outputs with shape (batch size, d_n, ... , d_1, output dim).

**Benefits:**
Allows more flexibility in the inputs that can be passed to TorchLayer and KerasLayer. [For example,](https://discuss.pennylane.ai/t/broadcast-keras-layer-to-sequence/759) sequence-to-sequence modelling is now possible.

**Possible Drawbacks:**
The proposed solution is recursive, which I understand is unpopular in Python, but this seemed the most intuitive approach. I can make it iterative easily enough if this is deemed to be unacceptable.

**Related GitHub Issues:**
#991

### CODE REVIEW

**Comment by josh146** [comment]
Thanks @DanielPolatajko for your PR! Let us know if you have any questions, or otherwise, simply leave a comment here when this PR is ready for review.

**Comment by DanielPolatajko** [comment]
Thanks @josh146. I think the PR is ready for review; I suppose the only question I have is if you know how can I fix the test check that failed? It was passing before I made a commit which did nothing other than edit the changelog, and I am unable to replicate the error locally.

**Comment by josh146** [comment]
Hi @DanielPolatajko, I've restarted the test suite, hopefully that should all pass in about 5-10 minutes!

File: tests/qnn/test_keras.py [hunk 1]
```diff
@@ -418,6 +418,25 @@ def c_default(w1, w2, w3, w4, w5, w6, w7, inputs=None):
...
+    @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(1))
+    @pytest.mark.parametrize("batch_size", [2])
+    @pytest.mark.parametrize("middle_dim", [5])
+    def test_call_broadcast(self, get_circuit, output_dim, middle_dim, batch_size, n_qubits):
+        """Test if the call() method performs correctly when the inputs argument is an arbitrary shape (that can
+        correctly be broadcast over), i.e., for input of shape (batch_size, dn, ... , d0) it outputs with shape
+        (batch_size, dn, ... , d1, output_dim)"""
+        c, w = get_circuit
+
+
```
Comment (trbromley) [inline tests/qnn/test_keras.py:430]:
```suggestion
```

File: tests/qnn/test_keras.py [hunk 2]
```diff
@@ -418,6 +418,25 @@ def c_default(w1, w2, w3, w4, w5, w6, w7, inputs=None):
...
+        layer = KerasLayer(c, w, output_dim)
+        x = tf.ones((batch_size, middle_dim, n_qubits))
+
+        layer_out = layer(x)
+
+        assert layer_out.shape == (batch_size, middle_dim, output_dim)
+
+
+
```
Comment (trbromley) [inline tests/qnn/test_keras.py:439]:
```suggestion
```

File: tests/qnn/test_keras.py [hunk 3]
```diff
@@ -418,6 +418,25 @@ def c_default(w1, w2, w3, w4, w5, w6, w7, inputs=None):
...
+        correctly be broadcast over), i.e., for input of shape (batch_size, dn, ... , d0) it outputs with shape
+        (batch_size, dn, ... , d1, output_dim)"""
+        c, w = get_circuit
+
+
+        layer = KerasLayer(c, w, output_dim)
+        x = tf.ones((batch_size, middle_dim, n_qubits))
+
+        layer_out = layer(x)
+
```
Comment (trbromley) [inline tests/qnn/test_keras.py:435]: Either in this test or a separate one, it would be good to check if the gradient is still accessible (e.g., even if we don't know the expected value, just to check it can be calculated and is not `None`).

Comment (trbromley) [reply tests/qnn/test_keras.py:435]: (The same suggestion applies to the torch test)

**Review by trbromley — APPROVED** [review]
Thanks @DanielPolatajko for the contribution! :100: I've left a few suggestions but overall approved.

One general question I had was how this would work if the QNode inputs or outputs were not flat? For non-flat QNode outputs (see the discussion [here](https://github.com/PennyLaneAI/pennylane/issues/936)), I don't see any problem. It's a bit more tricky though with non-flat inputs? I guess we'd need to adapt the `if len(tf.shape(inputs)) > 1` line, but currently we'd need a user to additional specify the input shape/number of dimensions.

Anyway, this is not a problem for now, just thinking in terms of future.

File: tests/qnn/test_keras.py (hunk 3 above)
Comment (DanielPolatajko) [reply tests/qnn/test_keras.py:438]: Good idea, have done that now

**Comment by DanielPolatajko** [comment]
> Thanks @DanielPolatajko for the contribution! 💯 I've left a few suggestions but overall approved.
> 
> One general question I had was how this would work if the QNode inputs or outputs were not flat? For non-flat QNode outputs (see the discussion [here](https://github.com/PennyLaneAI/pennylane/issues/936)), I don't see any problem. It's a bit more tricky though with non-flat inputs? I guess we'd need to adapt the `if len(tf.shape(inputs)) > 1` line, but currently we'd need a user to additional specify the input shape/number of dimensions.
> 
> Anyway, this is not a problem for now, just thinking in terms of future.

Cheers for that @trbromley, I've added all of your suggestions now. As far as having non-flat inputs, I think you're right that the user would have to specify the dimensions along which individual data points are found. I can't really see any other way to infer the dimensionality of individual inputs. Like you said though, if the user does provide that information, then its a pretty simple one-line fix, although it does bring up the point that there is an implicit assumption in this solution that the data points are in the last dimension of the input tensor. For multi-dimensional inputs this might not always be a given (I'm thinking of RGB channels in an image, for example), but this could easily be solved by reshaping the inputs to put the data points in the last few dimensions. As for outputs, I agree that its not a problem.

I'm still quite new to Pennylane and QML so I may well be wrong in saying this, but wouldn't non-flat inputs have to be flattened anyway before being passed through a QNode? So it might be a reasonable workaround to simply flatten the inputs along the relevant dimensions before passing it to the KerasLayer or TorchLayer.

**Comment by kessler-frost** [comment]
Hi @DanielPolatajko,

> but wouldn't non-flat inputs have to be flattened anyway before being passed through a QNode? So it might be a reasonable workaround to simply flatten the inputs along the relevant dimensions before passing it to the KerasLayer or TorchLayer.

My question is why would we have to flatten the input? As I understand it, Pennylane recently added support for density matrix allowing the QNode to return a non-flat output.

**Comment by DanielPolatajko** [comment]
Hi @kessler-frost, 
> My question is why would we have to flatten the input before passing it through the QNode? As I understand it, Pennylane recently added support for density matrix allowing the QNode to return a non-flat output.

I guess your question is the part I'm not 100% certain on, so I may be wrong, but I'll try to explain my thoughts a little more. I think maybe flatten was a misleading word choice.  What I'm getting at is that embeddings of classical data into quantum circuits usually operate (as far as I've been able to figure out) on vectors, so if you have some inputs in matrix form (for example, images), or even higher dimensional forms, they probably need to be somehow represented as vectors before being passed through the circuit. For the example of images, the quanvolution is one way of doing this, just directly flattening the input might be another.

As far as allowing QNodes to output density matrices is concerned, I don't think this is directly related to the input shape. I would have thought it would be possible to trace out one or more wires to obtain a density matrix, or to just make measurements on each wire, regardless of the particular inputs to the circuit.

**Comment by kessler-frost** [comment]
Ah, got it. I mentioned the density matrix part since I (wrongly?) thought single-dimensional (flattened) inputs will be required to get single-dimensional outputs from QNode, and thereby assuming that being the reason for flattening the input to QNode. I guess I will have to tinker more with QNode, image data, and the likes to form a better understanding of how all of this operates. Looking forward to the merging of this PR to continue experimenting.

File: .github/CHANGELOG.md [hunk 4]
```diff
@@ -16,7 +45,7 @@
 
 <h3>Contributors</h3>
 
-This release contains contributions from (in alphabetical order):
+This release contains contributions from (in alphabetical order): Daniel Polatajko
 
 Thomas Bromley
```
Comment (antalszava) [inline .github/CHANGELOG.md:50]:
```suggestion
This release contains contributions from (in alphabetical order):

Thomas Bromley, Daniel Polatajko
```

File: .github/CHANGELOG.md [hunk 5]
```diff
@@ -8,6 +8,35 @@
...
+
+  out = qlayer(x)
+  
+  print(out.shape)
+  ```
+
+  The output tensor has the following shape:
+
+  >>> out.shape
+  (5, 4, 4)
```
Comment (antalszava) [inline .github/CHANGELOG.md:39]: This will need to become, otherwise `>>>` is interpreted as three nested block quotes.
<pre>```pycon
>>> out.shape
(5, 4, 4)
```</pre>

File: tests/qnn/test_keras.py [hunk 6]
```diff
@@ -418,6 +418,27 @@ def c_default(w1, w2, w3, w4, w5, w6, w7, inputs=None):
         assert layer_out.shape == (batch_size, output_dim)
         assert np.allclose(layer_out[0], c(x[0], *weights))
 
+    @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(1))
+    @pytest.mark.parametrize("batch_size", [2])
+    @pytest.mark.parametrize("middle_dim", [5])
+    def test_call_broadcast(self, get_circuit, output_dim, middle_dim, batch_size, n_qubits):
+        """Test if the call() method performs correctly when the inputs argument is an arbitrary shape (that can
```
Comment (antalszava) [inline tests/qnn/test_keras.py:425]:
```suggestion
        """Test if the call() method performs correctly when the inputs argument has an arbitrary shape (that can
```

File: tests/qnn/test_keras.py [hunk 7]
```diff
@@ -418,6 +418,27 @@ def c_default(w1, w2, w3, w4, w5, w6, w7, inputs=None):
         assert layer_out.shape == (batch_size, output_dim)
         assert np.allclose(layer_out[0], c(x[0], *weights))
 
+    @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(1))
```
Comment (antalszava) [inline tests/qnn/test_keras.py:421]: How about testing for multiple qubits too? Also, we could maybe have 1-2 more values for `batch_size` and `middle_dim` (parameterization was already nicely set up :slightly_smiling_face: ). This latter point applies in the Torch case too.

File: .github/CHANGELOG.md [hunk 8]
```diff
@@ -8,6 +8,35 @@
 
 <h3>Improvements</h3>
 
+<h4>Broadcasting over multi-dimensional inputs to KerasLayer and TorchLayer</h4>
+
+* Most layers in Pytorch or Keras accept arbitrary dimension inputs, where each dimension barring the last (in the case where the actual weight function of the layer operates     on one-dimensional vectors) is broadcast over. This is now also supported by KerasLayer and TorchLayer [(#1062)](https://github.com/PennyLaneAI/pennylane/pull/1062).
```
Comment (antalszava) [inline .github/CHANGELOG.md:13]:
```suggestion
* Most layers in Pytorch or Keras accept arbitrary dimension inputs, where each dimension barring
  the last (in the case where the actual weight function of the layer operates on one-dimensional 
  vectors) is broadcast over. This is now also supported by KerasLayer and TorchLayer.
  [(#1062)](https://github.com/PennyLaneAI/pennylane/pull/1062).
```

**Review by antalszava — COMMENTED** [review]
This looks like a great addition @DanielPolatajko :1st_place_medal: :blush:

Left a couple of comments/suggestions, nothing major though. Would just go for a final skim through before approval.

File: tests/qnn/test_keras.py (hunk 7 above)
Comment (DanielPolatajko) [reply tests/qnn/test_keras.py]: Sure, sounds good. I added a couple extra values to each parameter by just picking a couple of small numbers, is there an established convention for choosing parameters for tests you'd prefer I followed?

Comment (antalszava) [reply tests/qnn/test_keras.py]: No, this will be great! Thanks!

**Review by antalszava — APPROVED** [review]
Looks good to me! 💯

**Comment by antalszava** [comment]
Thanks @DanielPolatajko again!
